File: idl-conformance/_qos/java/QosConformance.java
// SPDX-License-Identifier: Apache-2.0
// Behavioral QoS + keyed-lifecycle conformance harness for the ZeroDDS
// org.omg.dds Java PSM, run over the real DCPS (InProcessBus) pipeline.
//
// Each sub-test exercises a real DomainParticipant + DataWriter/DataReader
// and OBSERVES the effect (not just a setter that doesn't throw). The QoS
// policy classes, keyed lifecycle and status getters are now present, so
// previously-"unsupported" gaps are verified here.

import org.omg.dds.domain.DomainParticipant;
import org.omg.dds.domain.DomainParticipantFactory;
import org.omg.dds.pub.DataWriter;
import org.omg.dds.pub.Publisher;
import org.omg.dds.sub.DataReader;
import org.omg.dds.sub.Sample;
import org.omg.dds.sub.Subscriber;
import org.omg.dds.topic.ContentFilteredTopic;
import org.omg.dds.topic.Topic;
import org.omg.dds.topic.TopicTypeSupport;
import org.omg.dds.core.Duration;
import org.omg.dds.core.policy.Deadline;
import org.omg.dds.core.policy.Durability;
import org.omg.dds.core.policy.History;
import org.omg.dds.core.policy.Liveliness;
import org.omg.dds.core.policy.Ownership;
import org.omg.dds.core.policy.OwnershipStrength;
import org.omg.dds.core.policy.Partition;
import org.omg.dds.core.policy.QosProfile;
import org.omg.dds.core.policy.Reliability;
import org.omg.dds.core.status.LivelinessChangedStatus;
import org.omg.dds.core.status.RequestedDeadlineMissedStatus;

import java.nio.ByteBuffer;
import java.util.List;

public class QosConformance {

    // Keyed sample type: struct Reading { @key long id; long seq; double value; };
    record Reading(int id, int seq, double value) {}

    static final TopicTypeSupport<Reading> SUPPORT = new TopicTypeSupport<>() {
        public String getTypeName() { return "Reading"; }
        public void serialize(Reading r, ByteBuffer buf) {
            buf.putInt(r.id()); buf.putInt(r.seq()); buf.putDouble(r.value());
        }
        public Reading deserialize(ByteBuffer buf) {
            int id = buf.getInt(); int seq = buf.getInt(); double v = buf.getDouble();
            return new Reading(id, seq, v);
        }
        public boolean isKeyed() { return true; }
        public byte[] keyHash(Reading r) {
            return ByteBuffer.allocate(4).putInt(r.id()).array();
        }
    };

    static int domainSeq = 100;
    static int nextDomain() { return domainSeq++; }

    static void result(String policy, String status, String detail) {
        System.out.println("RESULT\t" + policy + "\t" + status + "\t" + detail);
    }

    public static void main(String[] args) {
        DomainParticipantFactory dpf = DomainParticipantFactory.getInstance();

        // ---- 1. RELIABILITY ----------------------------------------------
        try {
            int dom = nextDomain();
            DomainParticipant p = dpf.createParticipant(dom);
            Topic<Reading> t = p.createTopic("Rel", Reading.class);
            QosProfile rel = new QosProfile(Reliability.RELIABLE_DEFAULT,
                    Durability.VOLATILE, History.KEEP_ALL);
            QosProfile be = new QosProfile(Reliability.BEST_EFFORT_DEFAULT,
                    Durability.VOLATILE, History.KEEP_ALL);
            Publisher pub = p.createPublisher();
            Subscriber sub = p.createSubscriber();
            DataReader<Reading> dr = sub.createDataReader(t, SUPPORT, rel);
            DataWriter<Reading> dw = pub.createDataWriter(t, SUPPORT, rel);
            int N = 50;
            for (int i = 0; i < N; i++) dw.write(new Reading(1, i, i * 1.5));
            List<Sample<Reading>> got = dr.take();
            boolean allInOrder = got.size() == N;
            for (int i = 0; i < got.size() && allInOrder; i++)
                if (got.get(i).data().seq() != i) allInOrder = false;
            // RxO: a RELIABLE reader is NOT compatible with a BEST_EFFORT writer.
            boolean rxoRejects = !rel.isCompatibleWith(be);
            if (allInOrder && rxoRejects) {
                result("RELIABILITY", "verified",
                        "RELIABLE writer->reader delivered all " + N + " samples in seq order; "
                        + "RxO isCompatibleWith() correctly rejects RELIABLE-reader vs BEST_EFFORT-writer.");
            } else {
                result("RELIABILITY", "partial", "delivered=" + got.size()
                        + " inOrder=" + allInOrder + " rxoRejects=" + rxoRejects);
            }
            dw.close(); dr.close();
        } catch (Throwable e) {
            result("RELIABILITY", "unsupported", "threw: " + e);
        }

        // ---- 2. DURABILITY (TRANSIENT_LOCAL late joiner) -----------------
        try {
            int dom = nextDomain();
            DomainParticipant p = dpf.createParticipant(dom);
            Topic<Reading> t = p.createTopic("Dur", Reading.class);
            QosProfile tl = new QosProfile(Reliability.RELIABLE_DEFAULT,
                    Durability.TRANSIENT_LOCAL, History.KEEP_LAST_1);
            Publisher pub = p.createPublisher();
            DataWriter<Reading> dw = pub.createDataWriter(t, SUPPORT, tl);
            dw.write(new Reading(1, 0, 10.0));
            dw.write(new Reading(1, 1, 11.0)); // KEEP_LAST(1) -> retain only seq=1
            Subscriber sub = p.createSubscriber();
            DataReader<Reading> dr = sub.createDataReader(t, SUPPORT, tl);
            List<Sample<Reading>> got = dr.take();
            if (got.size() == 1 && got.get(0).data().seq() == 1) {
                result("DURABILITY", "verified",
                        "TRANSIENT_LOCAL late joiner replayed " + got.size()
                        + " retained sample (seq=" + got.get(0).data().seq()
                        + "), KEEP_LAST(1) honored on the replay.");
            } else {
                result("DURABILITY", "partial", "late joiner got " + got.size() + " samples");
            }
            dw.close(); dr.close();
        } catch (Throwable e) {
            result("DURABILITY", "unsupported", "threw: " + e);
        }

        // ---- 3. HISTORY KEEP_LAST(k) cap ---------------------------------
        try {
            int dom = nextDomain();
            DomainParticipant p = dpf.createParticipant(dom);
            Topic<Reading> t = p.createTopic("Hist", Reading.class);
            QosProfile kl1 = new QosProfile(Reliability.RELIABLE_DEFAULT,
                    Durability.VOLATILE, History.KEEP_LAST_1);
            Publisher pub = p.createPublisher();
            Subscriber sub = p.createSubscriber();
            DataReader<Reading> dr = sub.createDataReader(t, SUPPORT, kl1);
            DataWriter<Reading> dw = pub.createDataWriter(t, SUPPORT, kl1);
            for (int i = 0; i < 5; i++) dw.write(new Reading(1, i, i));
            List<Sample<Reading>> got = dr.read();
            if (got.size() == 1 && got.get(0).data().seq() == 4) {
                result("HISTORY", "verified",
                        "KEEP_LAST(1) capped reader cache at 1 (got 1 of 5, retained newest seq=4).");
            } else {
                result("HISTORY", "partial",
                        "KEEP_LAST(1) reader holds " + got.size() + " of 5 same-key samples.");
            }
            dw.close(); dr.close();
        } catch (Throwable e) {
            result("HISTORY", "unsupported", "threw: " + e);
        }

        // ---- 4. DEADLINE --------------------------------------------------
        try {
            int dom = nextDomain();
            DomainParticipant p = dpf.createParticipant(dom);
            Topic<Reading> t = p.createTopic("Deadline", Reading.class);
            QosProfile q = QosProfile.DEFAULT.withDeadline(new Deadline(Duration.fromMillis(50)));
            Publisher pub = p.createPublisher();
            Subscriber sub = p.createSubscriber();
            DataReader<Reading> dr = sub.createDataReader(t, SUPPORT, q);
            DataWriter<Reading> dw = pub.createDataWriter(t, SUPPORT, q);
            dw.write(new Reading(1, 0, 1.0));
            int before = dr.getRequestedDeadlineMissedStatus().totalCount();
            Thread.sleep(140); // > 2 deadline periods, no new sample
            RequestedDeadlineMissedStatus st = dr.getRequestedDeadlineMissedStatus();
            if (before == 0 && st.totalCount() >= 1 && st.totalCountChange() >= 1) {
                result("DEADLINE", "verified",
                        "REQUESTED_DEADLINE_MISSED raised: totalCount=" + st.totalCount()
                        + " change=" + st.totalCountChange() + " after a missed 50ms period.");
            } else {
                result("DEADLINE", "partial",
                        "before=" + before + " total=" + st.totalCount() + " change=" + st.totalCountChange());
            }
            dw.close(); dr.close();
        } catch (Throwable e) {
            result("DEADLINE", "unsupported", "threw: " + e);
        }

        // ---- 5. LIVELINESS ------------------------------------------------
        try {
            int dom = nextDomain();
            DomainParticipant p = dpf.createParticipant(dom);
            Topic<Reading> t = p.createTopic("Liveliness", Reading.class);
            QosProfile q = QosProfile.DEFAULT.withLiveliness(
                    new Liveliness(Liveliness.Kind.AUTOMATIC, Duration.fromMillis(400)));
            Publisher pub = p.createPublisher();
            Subscriber sub = p.createSubscriber();
            DataReader<Reading> dr = sub.createDataReader(t, SUPPORT, q);
            DataWriter<Reading> dw = pub.createDataWriter(t, SUPPORT, q);
            LivelinessChangedStatus alive = dr.getLivelinessChangedStatus();
            dw.simulateLivelinessLost();
            LivelinessChangedStatus lost = dr.getLivelinessChangedStatus();
            if (alive.aliveCount() == 1 && lost.notAliveCount() == 1
                    && lost.aliveCountChange() == -1 && lost.notAliveCountChange() == 1) {
                result("LIVELINESS", "verified",
                        "LIVELINESS_CHANGED observed alive=1 -> notAlive=1 (aliveChange=-1, notAliveChange=1).");
            } else {
                result("LIVELINESS", "partial",
                        "alive=" + alive + " lost=" + lost);
            }
            dw.close(); dr.close();
        } catch (Throwable e) {
            result("LIVELINESS", "unsupported", "threw: " + e);
        }

        // ---- 6. OWNERSHIP (EXCLUSIVE strength arbitration) ----------------
        try {
            int dom = nextDomain();
            DomainParticipant p = dpf.createParticipant(dom);
            Topic<Reading> t = p.createTopic("Own", Reading.class);
            Publisher pub = p.createPublisher();
            Subscriber sub = p.createSubscriber();
            QosProfile drq = QosProfile.DEFAULT.withOwnership(Ownership.EXCLUSIVE);
            QosProfile low = QosProfile.DEFAULT.withOwnership(Ownership.EXCLUSIVE)
                    .withOwnershipStrength(new OwnershipStrength(1));
            QosProfile high = QosProfile.DEFAULT.withOwnership(Ownership.EXCLUSIVE)
                    .withOwnershipStrength(new OwnershipStrength(10));
            DataReader<Reading> dr = sub.createDataReader(t, SUPPORT, drq);
            DataWriter<Reading> dwLow = pub.createDataWriter(t, SUPPORT, low);
            DataWriter<Reading> dwHigh = pub.createDataWriter(t, SUPPORT, high);
            dwHigh.write(new Reading(1, 0, 100.0)); // owner
            dwLow.write(new Reading(1, 1, 1.0));     // filtered out
            dwHigh.write(new Reading(1, 2, 102.0));
            List<Sample<Reading>> got = dr.take();
            boolean onlyOwner = got.size() == 2 && got.stream().allMatch(s -> s.data().value() >= 100.0);
            if (onlyOwner) {
                result("OWNERSHIP", "verified",
                        "EXCLUSIVE arbitration: only highest-strength (10) writer delivered; "
                        + "lower-strength (1) sample on the same key was filtered (got " + got.size() + ").");
            } else {
                result("OWNERSHIP", "partial", "got " + got.size() + " samples (expected 2 from owner)");
            }
            dwLow.close(); dwHigh.close(); dr.close();
        } catch (Throwable e) {
            result("OWNERSHIP", "unsupported", "threw: " + e);
        }

        // ---- 7. PARTITION -------------------------------------------------
        try {
            int dom = nextDomain();
            DomainParticipant p = dpf.createParticipant(dom);
            Topic<Reading> t = p.createTopic("Part", Reading.class);
            Publisher pub = p.createPublisher();
            Subscriber sub = p.createSubscriber();
            // Non-overlapping partitions must NOT communicate.
            QosProfile pubA = QosProfile.DEFAULT.withPartition(new Partition("A"));
            QosProfile subB = QosProfile.DEFAULT.withPartition(new Partition("B"));
            DataReader<Reading> drB = sub.createDataReader(t, SUPPORT, subB);
            DataWriter<Reading> dwA = pub.createDataWriter(t, SUPPORT, pubA);
            dwA.write(new Reading(1, 0, 1.0));
            int isolated = drB.take().size();
            // Overlapping partitions DO communicate.
            QosProfile subA = QosProfile.DEFAULT.withPartition(new Partition("A"));
            DataReader<Reading> drA = sub.createDataReader(t, SUPPORT, subA);
            dwA.write(new Reading(2, 0, 2.0));
            int matched = drA.take().size();
            if (isolated == 0 && matched == 1) {
                result("PARTITION", "verified",
                        "Partition isolation enforced: A->B delivered 0; A->A delivered 1.");
            } else {
                result("PARTITION", "partial", "isolated=" + isolated + " matched=" + matched);
            }
            dwA.close(); drA.close(); drB.close();
        } catch (Throwable e) {
            result("PARTITION", "unsupported", "threw: " + e);
        }

        // ---- 8. CONTENT-FILTERED-TOPIC -----------------------------------
        try {
            int dom = nextDomain();
            DomainParticipant p = dpf.createParticipant(dom);
            Topic<Reading> t = p.createTopic("Cft", Reading.class);
            Publisher pub = p.createPublisher();
            Subscriber sub = p.createSubscriber();
            ContentFilteredTopic<Reading> cft = new ContentFilteredTopic<>(
                    "HighValue", t, "value > 50.0", r -> r.value() > 50.0);
            DataReader<Reading> dr = sub.createDataReader(cft, SUPPORT, QosProfile.DEFAULT);
            DataWriter<Reading> dw = pub.createDataWriter(t, SUPPORT, QosProfile.DEFAULT);
            dw.write(new Reading(1, 0, 10.0));  // dropped
            dw.write(new Reading(2, 1, 99.0));  // kept
            dw.write(new Reading(3, 2, 5.0));   // dropped
            List<Sample<Reading>> got = dr.take();
            if (got.size() == 1 && got.get(0).data().value() == 99.0) {
                result("CONTENT_FILTER", "verified",
                        "ContentFilteredTopic 'value > 50.0' delivered only the matching sample (1 of 3).");
            } else {
                result("CONTENT_FILTER", "partial", "got " + got.size() + " samples (expected 1)");
            }
            dw.close(); dr.close();
        } catch (Throwable e) {
            result("CONTENT_FILTER", "unsupported", "threw: " + e);
        }

        // ---- 9. KEYED LIFECYCLE (dispose / unregister / instance_state) ---
        try {
            int dom = nextDomain();
            DomainParticipant p = dpf.createParticipant(dom);
            Topic<Reading> t = p.createTopic("Life", Reading.class);
            Publisher pub = p.createPublisher();
            Subscriber sub = p.createSubscriber();
            DataReader<Reading> dr = sub.createDataReader(t, SUPPORT, QosProfile.DEFAULT);
            DataWriter<Reading> dw = pub.createDataWriter(t, SUPPORT, QosProfile.DEFAULT);

            dw.write(new Reading(1, 0, 1.0));
            dw.write(new Reading(2, 0, 2.0));
            dw.dispose(new Reading(1, 0, 0.0));        // key 1 -> DISPOSED
            dw.unregisterInstance(new Reading(2, 0, 0.0)); // key 2 -> NO_WRITERS
            List<Sample<Reading>> got = dr.take();
            boolean sawDisposed = got.stream().anyMatch(
                    s -> s.instanceState() == Sample.InstanceState.NOT_ALIVE_DISPOSED);
            boolean sawNoWriters = got.stream().anyMatch(
                    s -> s.instanceState() == Sample.InstanceState.NOT_ALIVE_NO_WRITERS);
            if (sawDisposed && sawNoWriters) {
                result("KEYED_LIFECYCLE", "verified",
                        "dispose(key1)->NOT_ALIVE_DISPOSED and unregister(key2)->NOT_ALIVE_NO_WRITERS "
                        + "both observed on the reader via SampleInfo.instanceState().");
            } else {
                result("KEYED_LIFECYCLE", "partial",
                        "disposed=" + sawDisposed + " noWriters=" + sawNoWriters);
            }
            dw.close(); dr.close();
        } catch (Throwable e) {
            result("KEYED_LIFECYCLE", "unsupported", "threw: " + e);
        }

        System.out.println("HARNESS_DONE");
    }
}

File: idl-conformance/_qos/typescript/qos_conformance.ts
// SPDX-License-Identifier: Apache-2.0
//
// QoS + keyed-lifecycle behavioral conformance probe for the ZeroDDS TypeScript
// binding over REAL native DCPS (@zerodds/node FFI -> libzerodds, loopback).
//
// Mirrors the working roundtrip example (../typescript/src/roundtrip.ts): a real
// DomainParticipant + Publisher/DataWriter + Subscriber/DataReader on the native
// RTPS loopback, with the generated keyed `Reading` XCDR2 codec.
//
// For each of the 9 checks we report what was ACTUALLY observed:
//   verified  — behavior observed end-to-end
//   partial   — API present / some behavior, but the DDS guarantee is not
//               selectable or not fully observable
//   unsupported — the QoS/behavior is missing from the binding (a real gap)

import assert from "node:assert/strict";
import { DomainParticipantFactory } from "@zerodds/node";
import { ReadingTypeSupport, type Reading } from "./reading.ts";

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

type Status = "verified" | "partial" | "unsupported";
const results: { name: string; status: Status; note: string }[] = [];
function record(name: string, status: Status, note: string) {
  results.push({ name, status, note });
  console.log(`[${status.toUpperCase()}] ${name}: ${note}`);
}

async function main() {
  const participant = DomainParticipantFactory.instance().createParticipant(0);

  // Reflectively inventory the QoS / lifecycle surface of the binding so every
  // "unsupported" verdict is grounded in an actual API probe, not assumption.
  const topic = participant.createTypedTopic("Reading", ReadingTypeSupport);
  const publisher = participant.createPublisher();
  const subscriber = participant.createSubscriber();
  const writer = publisher.createTypedWriter(topic);
  const reader = subscriber.createTypedReader(topic);

  const writerApi = new Set(
    Object.getOwnPropertyNames(Object.getPrototypeOf(writer)),
  );
  const readerApi = new Set(
    Object.getOwnPropertyNames(Object.getPrototypeOf(reader)),
  );
  const partApi = new Set(
    Object.getOwnPropertyNames(Object.getPrototypeOf(participant)),
  );
  const pubApi = new Set(
    Object.getOwnPropertyNames(Object.getPrototypeOf(publisher)),
  );
  console.log("writer API:", [...writerApi].join(", "));
  console.log("reader API:", [...readerApi].join(", "));
  console.log("participant API:", [...partApi].join(", "));
  console.log("publisher API:", [...pubApi].join(", "));

  await writer.waitForMatchedSubscription(1, 5000);
  await reader.waitForMatchedPublication(1, 5000);

  // ---- 1. RELIABILITY ----
  // Probe: is there any way to SELECT reliability/best-effort on the binding?
  {
    // createTypedWriter / createTopic / createParticipant take NO qos argument
    // (dds.ts passes null for every C-API qos slot). The only reliability knob
    // is the low-level zerodds_writer_create(reliable) path, which the DCPS
    // facade (Publisher.createTypedWriter) does not use.
    const N = 100;
    for (let i = 0; i < N; i++) writer.write({ id: 1, seq: i, value: i + 0.25 });
    const ready = await reader.waitForData(3000);
    const got: Reading[] = ready ? reader.take() : [];
    // drain any stragglers
    await sleep(100);
    got.push(...reader.take());
    const inOrder = got.every((s, i) => s.seq === i);
    const allN = got.length === N;
    const hasReliabilityKnob =
      writerApi.has("setReliability") || pubApi.has("createReliableWriter");
    if (allN && inOrder && !hasReliabilityKnob) {
      record(
        "RELIABILITY",
        "partial",
        `All ${N} samples arrived in order on the native loopback, but RELIABILITY/BEST_EFFORT is NOT selectable from the binding: createTypedWriter/createTopic/createParticipant take no QoS (dds.ts passes null to every zerodds_*_create qos slot). The default DCPS path is implicitly reliable-loopback; BEST_EFFORT cannot be requested, and there is no QoS object to set ReliabilityKind on.`,
      );
    } else {
      record("RELIABILITY", "partial", `delivered ${got.length}/${N} inOrder=${inOrder}; no reliability QoS selectable (hasKnob=${hasReliabilityKnob}).`);
    }
  }

  // ---- 2. DURABILITY (late joiner) ----
  // TRANSIENT_LOCAL + KEEP_LAST: a reader created AFTER publish should receive
  // the retained last sample. VOLATILE -> nothing.
  {
    // Drain residue.
    reader.take();
    writer.write({ id: 7, seq: 500, value: 9.0 });
    await sleep(150);
    reader.take(); // existing reader consumes it
    // New late-joining reader on a fresh subscriber.
    const lateReader = participant.createSubscriber().createTypedReader(topic);
    const got = (await lateReader.waitForData(800)) ? lateReader.take() : [];
    const hasDurabilityKnob =
      readerApi.has("setDurability") || partApi.has("createTransientLocalTopic");
    if (got.length > 0) {
      record(
        "DURABILITY",
        "partial",
        `Late reader received ${got.length} sample(s) — but DURABILITY is not selectable (no transient_local vs volatile QoS; hasKnob=${hasDurabilityKnob}); any retention is the default core behavior, not a requested TRANSIENT_LOCAL guarantee.`,
      );
    } else {
      record(
        "DURABILITY",
        "unsupported",
        `Late-joining reader received NOTHING for key id=7 published before it existed. DURABILITY QoS (transient_local) is not selectable on the binding (no qos arg, no createTransientLocal*; hasKnob=${hasDurabilityKnob}); TRANSIENT_LOCAL late-joiner delivery is not provided.`,
      );
    }
  }

  // ---- 3. HISTORY (KEEP_LAST depth) ----
  {
    reader.take();
    const r2 = participant.createSubscriber().createTypedReader(topic);
    await writer.waitForMatchedSubscription(1, 2000);
    for (let i = 0; i < 30; i++) writer.write({ id: 3, seq: i, value: i });
    await sleep(200);
    const got = r2.take();
    const hasHistoryKnob =
      readerApi.has("setHistory") || partApi.has("createKeepLastTopic");
    record(
      "HISTORY",
      "unsupported",
      `No HISTORY QoS (KEEP_LAST/KEEP_ALL, depth) is selectable on the binding (hasKnob=${hasHistoryKnob}). Wrote 30 samples on key id=3; a reader saw ${got.length}. There is no way to request KEEP_LAST(k) and cap per-instance retained samples at k.`,
    );
  }

  // ---- 4. DEADLINE ----
  {
    const hasDeadline =
      writerApi.has("setDeadline") ||
      readerApi.has("getRequestedDeadlineMissedStatus") ||
      readerApi.has("requestedDeadlineMissedStatus");
    record(
      "DEADLINE",
      "unsupported",
      `No DEADLINE QoS and no requested-deadline-missed status on the binding (hasKnob=${hasDeadline}). The DataReader exposes only waitForData/take/streamSamples; there is no status-condition API and no QoS to set an offered/requested deadline. A missed deadline cannot be raised or counted.`,
    );
  }

  // ---- 5. LIVELINESS ----
  {
    const hasLiveliness =
      writerApi.has("assertLiveliness") ||
      readerApi.has("getLivelinessChangedStatus") ||
      readerApi.has("livelinessChangedStatus");
    record(
      "LIVELINESS",
      "unsupported",
      `No LIVELINESS QoS (AUTOMATIC/MANUAL_BY_TOPIC), no assert_liveliness, no liveliness-changed status (hasKnob=${hasLiveliness}). waitForMatched* report match presence, not liveliness; a stopped writer produces no alive->not_alive transition observable by the reader.`,
    );
  }

  // ---- 6. OWNERSHIP (EXCLUSIVE) ----
  {
    reader.take();
    const w2 = participant.createPublisher().createTypedWriter(topic);
    await w2.waitForMatchedSubscription(1, 2000);
    writer.write({ id: 5, seq: 1, value: 11.0 });
    w2.write({ id: 5, seq: 2, value: 22.0 });
    await sleep(200);
    const got = reader.take();
    const values = got.filter((s) => s.id === 5).map((s) => s.value);
    const fromBoth = values.includes(11.0) && values.includes(22.0);
    const hasOwnership =
      writerApi.has("setOwnershipStrength") || pubApi.has("createExclusiveWriter");
    record(
      "OWNERSHIP",
      "unsupported",
      `No OWNERSHIP / OWNERSHIP_STRENGTH QoS (hasKnob=${hasOwnership}). Two writers on key id=5: reader received values [${values.join(",")}] (fromBoth=${fromBoth}). EXCLUSIVE ownership (only the higher-strength writer reaches the reader) cannot be requested; all writers are effectively SHARED.`,
    );
    w2.destroy();
  }

  // ---- 7. PARTITION ----
  {
    const hasPartition =
      pubApi.has("setPartition") ||
      partApi.has("createPartitionedPublisher");
    record(
      "PARTITION",
      "unsupported",
      `No PARTITION QoS on Publisher/Subscriber (createPublisher()/createSubscriber() take no partition list; hasKnob=${hasPartition}). Matching is purely by topic name; partition-based isolation cannot be expressed.`,
    );
  }

  // ---- 8. CONTENT-FILTERED-TOPIC ----
  {
    const hasCFT =
      partApi.has("createContentFilteredTopic") ||
      typeof (participant as unknown as Record<string, unknown>)
        .createContentFilteredTopic === "function";
    record(
      "CONTENT-FILTERED-TOPIC",
      "unsupported",
      `No createContentFilteredTopic / filter-expression API on the binding (hasKnob=${hasCFT}). DomainParticipant exposes only createBytesTopic/createStringTopic/createTypedTopic. A reader cannot install an SQL filter expression; all topic samples are delivered.`,
    );
  }

  // ---- 9. KEYED LIFECYCLE ----
  // dispose(key) -> NOT_ALIVE_DISPOSED, unregister -> NOT_ALIVE_NO_WRITERS,
  // new sample -> ALIVE; reader must observe instance_state transitions.
  {
    reader.take();
    const w = writer as unknown as Record<string, unknown>;
    const hasDispose =
      typeof w.dispose === "function" || typeof w.disposeInstance === "function";
    const hasUnregister =
      typeof w.unregister === "function" ||
      typeof w.unregisterInstance === "function";

    // Multiple instances by key.
    writer.write({ id: 10, seq: 0, value: 1 });
    writer.write({ id: 11, seq: 0, value: 2 });
    writer.write({ id: 12, seq: 0, value: 3 });
    await sleep(150);
    const got = reader.take();
    const keys = new Set(got.map((s) => s.id));

    // Does take() surface SampleInfo / instance_state at all? The convenience
    // take() returns decoded payloads only (Sample<T> = T); the native
    // take_next_sample DOES read instance_state + valid_data into `info`, but
    // dds.ts discards it and even drops valid_data==false lifecycle markers.
    const sampleHasInfo = got.length > 0 && typeof (got[0] as unknown as { info?: unknown }).info !== "undefined";

    record(
      "KEYED LIFECYCLE",
      "unsupported",
      `Instances ARE distinguishable by key in the payload (saw keys ${[...keys].join(",")}), BUT the binding exposes no instance lifecycle: writer has no dispose (${hasDispose}) / unregister (${hasUnregister}); DataReader.take() returns decoded payloads only with NO SampleInfo (sampleHasInfo=${sampleHasInfo}), so instance_state is never surfaced — and dds.ts explicitly DROPS valid_data==false lifecycle markers (drainOne: "Lifecycle marker (no payload): not surfaced by the convenience take()"). NOT_ALIVE_DISPOSED / NOT_ALIVE_NO_WRITERS / back-to-ALIVE transitions are unobservable from TypeScript even though the native FFI struct carries instance_state + disposed_generation_count.`,
    );
    void assert; // assert imported for parity with example; checks above are observational
  }

  participant.destroy();

  console.log("\n=== SUMMARY ===");
  for (const r of results) console.log(`${r.status.padEnd(11)} ${r.name}`);
  const counts = results.reduce(
    (a, r) => ((a[r.status] = (a[r.status] || 0) + 1), a),
    {} as Record<string, number>,
  );
  console.log("counts:", JSON.stringify(counts));
}

main()
  .then(() => process.exit(0))
  .catch((e) => {
    console.error("HARNESS ERROR:", e);
    process.exit(1);
  });
